src/self-update/README.md

```diff
@@ -0,0 +1,89 @@
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Details about the SM Plugin
+
+* A designated Software Management module type - namely "tedge" - will be used manage the self-update.
+
+* The existing SM plugin "apt" will be used and extended to manage that module type "tedge". 
+  * A device owner can implement an own plugin for module type "tedge", e.g. to use any other package manager than APT for self-update. 
+
+* A softlink to be created in `/etc/tedge/sm-plugins/` named `tedge` that points to `apt` plugin, will make the SM agent aware.
+
+* The plugin managing module type "tedge" must take all packages of an update request at once from SM Agent (i.E. Plugin API command "update-list" must be used). That is to get the plugin independent from potential restarts of the SM agent (those restarts are part of updating the SM agent).
+
+* Before any package update for module type "tedge" starts, all needed installation resources for the entire update request must be downloaded (i.E. all \*.deb packages for APT). That is to get the plugin uncoupled from any potential network issue during update procedure.
+  * For all packages with some URL given in the update request the SM Agent manages download before plugin start, and provides local file paths to the plugin.<br/>
+    TODO: What happens if one package download fails in SM agent? Does plugin maybe need to check if all files are locally available?
+  * For all packages without any URL given in the update request the Package Manager (i.E. APT) will manage the download as part of the update. Therefor the plugin must instruct the package manager to first download **all** packages. 
+    * If all packages were successfully downloaded, the package manager can start updating packages.<br/>
+    * Instead, if one or more download fails the plugin must stop before any installed package was touched by update, and report the entire update process as failed to the SM agent<br/>
+
+* After processing the complete update request the plugin will publish the overall exit code as MQTT retain message to topic `tedge/plugins/software/<plugin name>`, and exits.
+
+* As both module types "tedge" and "apt" use same package manager APT, the plugin manages a blacklist to allow later to distinguish packages reported by Package Managers `apt list` command.
+  * The blacklist is a TOML file that stores package names of all packages that were installed/updated with module type "tedge".
```

Comment:
I would simply use the prefix `tedge`.

Reply:
My previous comment was not clear.

I don't see the point of using a blacklist file. I would simply tag as managed by `tedge` all the packages which name starts with `tedge`.

Furthermore, I think that the proposed approach with a blacklist file will not work. One starts with an empty file, so *all* the packages, including `tedge`, will be marked as of the `pat` type. The device admin will not change that type before sending a request and the `tedge` package will never be listed as special.

Reply:
I suppose this could be workaround with using a prefix for tedge packages, for drop 1 we can assume all packages that names start with `tedge_` are special cases. This would simplify the implementation.

Reply:
I have replaced the blacklist approach with naming convention (prefix "tedge_") as below:

- As the plugin serves both module types "tedge" and "apt", the SM Agent will execute the plugin twice with the LIST command; once for each mode type. To distinguish packages listed by APT package manager per module type all packages prefixed with "tedge_" are assumed as module type "tedge".
   - When the plugin is called with LIST command for module type "tedge", packages prefixed with "tedge_" are reported.
   - When the plugin is called with LIST command for module type "apt", all packages withoutprefix "tedge_" are reported.